Multiple parts

marchantpeter opened the pull request:
It can now generate talking scores for scores with multiple instruments / parts!  

Create midi files for the select instruments.  If an instrument has multiple parts - a midi file is created for each individual part and all parts together.  Option to create midi file for all instruments played together / just selected / just unselected instruments.  

Add a click track to the midi files and change the speed; whilst following time signature and tempo changes - hopefully!  

The midi files are now generated on demand as opposed to with the talking score due to the number of files ie 18 files for each segment of a piano score.  

Added options for pitch / rhythm names / dot position etc.

The description is improving - but still more to do.  

It can now describe beat fractions ie beat 1 - dotted crotchet, beat 2.5 quaver - as opposed to saying the quaver was on beat 2.  It still groups notes into the same beat - it doesn't treat semiquavers as beat 1, 1.25, 1.5, 1.75...

Imrpoved summarisation - includes time signatures / key signatures / tempo changes. 

Repetition is announced within the score eg "Bars 29 through 32 were first used at 1 and lately used at 9".   Ie the first time a section was used along with the most recent time.

Significant speed improvements - over twice as fast in some scores.  This speed increase was after we started generating midi files on demand ie the improvement is just in generating the talking score - not by simply skipping all the midi files!  

Files with the same hash can be uploaded and it will prepend "-ver2" - to "-ver9" on the end.  

Bug fixes.

marchantpeter committed:
Get the selected instruments from options.html into .opts json file and then into HTMLTalkingScoreFormatter

marchantpeter committed:
fix wrong variable typo causing division by 0.  Don't analyse Harmony objects as chords as it affects grace notes etc

marchantpeter committed:
get the part numbers for selected and unselected instrumnts - part_instruments={}

marchantpeter committed:
Play full score just all selected or all unselected parts together

marchantpeter committed:
Describe dotted rhythms

marchantpeter committed:
Call unnamed instruments 'instrument X (unnamed)' instead of None

marchantpeter committed:
Added options for pitch / rhythm / octave description and position and announcement.  Also option to colour text / background using Figure Notes colours

marchantpeter committed:
Add option for phonetic pitch names - charlie / delta / echo etc

marchantpeter committed:
Play midis for individual instruments.  Need to add each hand for instruments

marchantpeter committed:
play instruments and parts ie hands for instruments with 2 parts - for segments.  Needs tidying up / refactoring

marchantpeter committed:
temporary fix for bars that don't have any notes or rests not having a key in the dictionary

marchantpeter committed:
Improved descriptions on options page etc

marchantpeter committed:
remove original calls to score.generate_midi_for_part_range as midi files are currently being generated twice...

marchantpeter committed:
started just making links to midi files so they can be generated when needed by views.py

marchantpeter committed:
change allowed_hosts to work on talkingscores.org too

marchantpeter committed:
I think pitch_index was supposed to order description so that eg dynamics came first.  But it ordered by pitch which meant notes inside a beat would be ordered by pitch!  Temporary fix - Renamed to order_index and made dynamics 0 and notes 1.  TODO - dynamics are still described at the start of the beat - not attached to a note.

marchantpeter committed:
Start - generate a midi file on demand from views.py if it doesn't exist already - as opposed to generate all midis during talking score generation

marchantpeter committed:
Update google analytics to work with .org instead of .co.uk

marchantpeter committed:
working on creacting midi files when requested.  Works out which instruments and parts are needed.  Makes a midi file for an individual part

marchantpeter committed:
Generate midi on demand for instruments and parts.  Needs tidying up

marchantpeter committed:
Play full score - all / selected / unslected / instruments

marchantpeter committed:
Describe parts individually!  Add tempo and click dropdowns to change onclick() destination for play links.

marchantpeter committed:
Adjust tempo and started click track

marchantpeter committed:
Add click track.  Just Bars / beats - removed quavers / start of bar - to reduce number of midi files needed

marchantpeter committed:
Fix multiple voices.  element.id is a str and needed to be an int.

marchantpeter committed:
make midi paths use BASE_DIR for absolute paths instead of relative so it works on pythonAnywhere and not just on windows...

marchantpeter committed:
pickup bar click track full score almost works. Tempo offset not quite. Needs lots of tidying up

marchantpeter committed:
Add ofset to tempos after adding click track with pickup bar

marchantpeter committed:
Don't make midi files for single part instruments twice - ins and p.  Just use ins

marchantpeter committed:
Upgrade from music21 6.3.0 to 7.3.3 - so that percussion sounds work in midi files - for snare and hihat in click track.  Also should improve speed.

marchantpeter committed:
Auto summary works on multiple parts.  Started describe_repetition_summary() to not just list all bar usages!  Improved document outline for talking score.  Need to do other pages.

marchantpeter committed:
Refactored a bit of describe_repetition_summary and moved repeated code into functions.  Started general_summary

marchantpeter committed:
summarise repetition section lengths

marchantpeter committed:
Describe bar groups and single bars are repeated - at the right point in the score.  Need to add intervals and rhythm.

marchantpeter committed:
Added repetition in context for rhythm and intervals.  Refactored into functions to avoid repeating similar code.

marchantpeter committed:
Describe beat fractions ie beat 1 - dotted quarter note, beat 2.5 - eight note.  Comma between notes  in beat

marchantpeter committed:
Significantly improve speed of describing update_events_for_measure by using collect=('TimeSignature') - so it calculates the beat quicker.  Eg 46 bars of Pachellbel's Canon piano - 26.19s to 17.665

marchantpeter committed:
Improve speed of parsing spanners by only looking at current part up to current segment instead of entire score for each segment.

marchantpeter committed:
Improve speed from 18s to 10s for Pachelbel. Make a dictionary of TimeSignatures before creating the TalkingScore to avoid using Collect when getting Measures.

marchantpeter committed:
Upload a file with the same hash and ver2 to ver9 will be prepended to the file name

marchantpeter committed:
Rename .xml to .musicxml file after upload.  Add the notes to the talking score before the spanners - becuase the beats aren't sorted in order after creating the talking score.  There might have been a spanner in beat 2 but notes in beat 1...  Updated the change log.

marchantpeter committed:
capitalize the first letter of the summary and repetition

marchantpeter committed:
Add sitemap.xml

marchantpeter committed:
Summarise time / key / tempo changes.  Describe time / key changes in the score at the start of each bar.

marchantpeter committed:
Cludge for bars with nothing in.  Use tempo referent ie beat length.  Use options to describe tempo referent - beat description and dot position.

marchantpeter committed:
Some midi files were indexed.  Allow /midis in robots.txt - then add  x-tag-robots: noindex to http response of midi files.  Also grammar correction in MusicAnalyser

marchantpeter committed:
Add slight offset when inserting scaled tempo at start of stream to make sure it occurs after an existing tempo - if present

marchantpeter committed:
Fix crash with time_sig_index.  Cludge to make scores with non sequential bar numbers kind of work - but still miss some bars!

marchantpeter committed:
Describe voice 1 first - ie don't reverse the voice dictionary

bentimms commented on lib/talkingscoreslib.py:
```diff
@@ -14,8 +14,11 @@
 logger = logging.getLogger("TSScore")
+import time
+
+global settings
```
Is this required to access options from the Django settings?  did you already consider using an import rather than relying on global scope?

marchantpeter replied:
During HTMLTalkingScoreFormatter.__init__() - It reads in the options that the user set for this talking score and stores them in settings.  But later on eg when it is rendering each individual note - it needs to know whether to put the dot before or after the rhythm...  Or similar question for the octave description / tempo etc.  I did wonder if the fact that I was trying to make something global meant there was a better way to do it?  I could pass relevant settings down to the notes etc?

bentimms commented on lib/midiHandler.py:
```diff
@@ -0,0 +1,295 @@
+        bsi = int(self.queryString.get("bsi"))
+        self.selected_instruments = []
+        while (bsi>1):
+            print ("bsi = " + str(bsi))
```
Did you consider using your logger for all the print statements?

marchantpeter replied:
I like that the logger will output to the log file - which I use for errors etc.  

But I tend to use print to show much more information for debugging as I work - so the log file doesn't get too big.  
I can see them in the console while I am testing - which is useful for scores that take a while to process - I can see how far it is going - while waiting for the error...
I don't actually know where / if pythonanywhere outputs them to?  

Before I commit - I sometimes tidy up some of the print statements and sometimes keep some that I think could still be useful.  
But I gradually miss a few - until there are probably loads that I don't need anymore...  I should go back and tidy them up.  

Or are you saying I could use the logger to output to the console and it would be better than print statements?

marchantpeter replied:
Fixed for midiHandlery.py in faff2e5 and ccda6c6
logger.debug to output to the console instead of to the log file.
logger format adds line numbers etc too!

bentimms commented on lib/midiHandler.py:
```diff
@@ -0,0 +1,295 @@
+        for part_index, part in enumerate(self.score.flat.getInstruments()):
+            print ("part_index = " + str(part_index) )
+            print (part)
+            if part.partId!=prev_instrument:
```
I'd highly recommend installing a "linter" plugin such as autopep8 to let it handle formatting your code to a common standard.

bentimms replied:
If you do, just make sure you check in any formatting changes separately to substantive changes.

marchantpeter replied:
I have now installed autopep8.  

I wanted it to fix the formatting as I type - in settings.json
"editor.formatOnType": true,
But that didn't seem work.  

If I right click and choose format docuent - that works.
Yup - I made that mistake in another project - I wanted to format the few lines I'd written - but it did the entire file and my diff was incomprehensible!

However - if I right click and do Format Selection - I get a message saying "Configure default formatter.  Extension 'autopep8' is configured as formatter but it cannot format 'python -files"

marchantpeter replied:
Fixed in 74276ac
I couldn't get it to work on just selected text - but the whole file works fine.
It does make a big difference to how the code looks!
I increased the maximum line length - because it was putting comments after the line they refer to - whereas I like them on the line or on the line before.

bentimms commented on lib/midiHandler.py:
```diff
@@ -0,0 +1,295 @@
+                end=0
+        
+        for click in ['n', 'be']:
+            self.tempo_shift = 0 #in a pickup bar - a rest is added to the start of bar 0 to make it a full bar - this offset needs adding to all future tempos 
```
Consider taking out the middle of very long for loop blocks and turning it into a function.

marchantpeter replied:
Yes - there are some rather long sections there.  As long as I name the functions well - it will make the code easier to read!

marchantpeter replied:
Fixed in 79fe761
I initially moved parts out into separate functions - then decided the functions were quite similar - so made them just one and added extra parameters.  I need to get the balance between lots of similar functions and one really complicated function used in several places for slightly different purposes...

bentimms commented on lib/talkingscore.html:
```diff
@@ -5,6 +5,7 @@
     <meta http-equiv="content-type" content="text/html; charset=utf-8"/>
     <meta name="robots" content="noindex" />
     
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.2/jquery.min.js"></script>
```
I'd highly recommend taking a look at https://htmx.org for interactivity.  It plays very nicely with Django.

marchantpeter replied:
I probably don't need to use jquery here.  I think I am only using it to find document elements!  I shall have a look at htmx

marchantpeter replied:
Fixed in 3329eb4 and 6816528
Might have reduced compatibility with older browsers?
I didn't think adding HTMX seemed necessary for what we were currently doing with jQuery?

bentimms commented on lib/talkingscoreslib.py:
```diff
@@ -324,6 +553,114 @@ def group_chord_pitches_by_octave(self, chord):
+    def generate_midi_filename_sel(self, prefix, range_start=None, range_end=None, output_path="", sel=""):
+        base_filename = os.path.splitext(os.path.basename(self.filepath))[0]
+        if (range_start!=None):
+            midi_filename = os.path.join(output_path, ("%s.mid?sel=%s&start=%d&end=%d&t=100&c=n" % ( base_filename, sel, range_start,range_end )) )
```
Have a look at Python f-strings for this sort of thing https://realpython.com/python-f-strings/

marchantpeter replied:
Ah yes - I use something similar in React - it is much easier to read

marchantpeter replied:
Fixed in 5fceb81
Thank you - f-strings do greatly improves readability!

bentimms commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
 Django>=2.2.10
```
I believe this is quite an insecure version of Django now, so it's worth investigating how easy it would be to use the latest version.

marchantpeter replied:
I've updated music21 a few times - but even that is a year or so old now.  But yes - Django must be very out of date!

marchantpeter replied:
Fixed in a3e987b
Upgrading Django required upgrading Python - which meant upgrading everything else...
Possibly could have upgraded Python a bit further?
I was expecting lots of code to update for talkingscores - but very few changes were needed :-)

bentimms commented on talkingscoresapp/views.py:
```diff
@@ -85,9 +103,13 @@ def score(request, id, filename):
-    fr = FileResponse(open("staticfiles/data/" + id + "/" + filename, "rb"))
+    mh = MidiHandler(request.GET, id, filename)
+    midiname = mh.get_or_make_midi_file()
+    fr = FileResponse(open(BASE_DIR+"/staticfiles/data/" + id + "/" + midiname, "rb"))
```
Consider using os.path.join?

marchantpeter replied:
I think it was using os.path.join originally.  I'll have another look at that section.

marchantpeter replied:
Fixed in 76e0b40

bentimms reviewed (commented):
Hi Peter - I thought I'd try out a GitHub review.  I did a quick sweep through and made a couple of comments – not an exhaustive review, but any stretch of the imagination.  There's a lot of code in this pull request - a lot of work!

marchantpeter committed:
Describe tuplets.  Probably needs improving - but a start.

marchantpeter committed:
Update homepage to explain that user must use a screen reader to make the score talk...

marchantpeter committed:
sanitise uploaded file names and remove apostrophes - otherwise it goes to options page then back to homepage...

marchantpeter committed:
Fix crash before options screen if 1st bar has no key signature

marchantpeter committed:
Improve error handling
Show error page instead of crashing if a problem occurs before the options screen.
Send developer an email when the error page is shown - even if the user doesn't want to be notified when it is fixed

marchantpeter committed:
Remove jquery from base_site.html
Use standard javascript to select the active tab in the nav bar.
Alter css to make active tab stand out more and improve hover etc

marchantpeter committed:
Remove jQuery from the score
Updates the tempo and click lists and play links with standard javascript instead of jquery

marchantpeter committed:
Prepared logging to replace print
Added 2 handlers to the logger - one for the console and one for the file.
Anything at level DEBUG will only go to the console - with line numbers!
Can also use F strings in the logging!

marchantpeter committed:
Replaced all print statements with logger.debug in midiHandler.py

marchantpeter committed:
Linted with autopep8

marchantpeter committed:
refactor make_midi_files to move some similar code into another functions

marchantpeter committed:
Replace %s with f-string - greatly improves readability!

marchantpeter committed:
Use os.path.join instead of adding slashes manually

marchantpeter committed:
Upgrade django from 2.2.10 to 4.2.6
Required upgrade of python from 3.7 to 3.10
This kind of forced update of Jinja and music21
Seemed sensible to upgrade requests too
Pathvalidate was only 1 version behind - but updated it also

marchantpeter committed:
Fixed show repetition_in_contexts for correct part
Could sometimes cause crash from undefined index - as it was using index of selected parts but needed index of all parts

marchantpeter committed:
Removing jquery (previously) stopped bootstrap 3 mobile menu toggle working
upgraded to bootstrap 5 which doesn't need jquery
this meant bits of css and layout etc needed updating too

marchantpeter committed:
Fix crash if tempo.number can't be read
Sometimes tempo has soundingNumber but not number - so use soundingNumber
needed for tempo description and scaling tempo for midi files

marchantpeter committed:
Fix a crash in get_title.
In some musicxml files - the attributes we are looking for in text boxes might not exist - so use hasattr()

marchantpeter committed:
Don't crash due to Unpitched notes.
Describe them in the score.

marchantpeter committed:
Fix for scores with only 1 bar
summarise repetition was wrong - and isn't needed
Fix description and playing

marchantpeter committed:
Fix crash when a score doesn't have a time signature

marchantpeter committed:
Update change log page
Make Notify Me button more visible on Error page
